### ISSUE TITLE
XCompile-Android.txt not working (snprintf missing)

### ISSUE THREAD

**Issue by LordOfDragons** [kcat/openal-soft#274]
Build OpenAL
-- The C compiler identification is GNU 4.9.0
-- The CXX compiler identification is GNU 4.9.0
-- Check for working C compiler: /opt/android-ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-gcc
-- Check for working C compiler: /opt/android-ndk/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64/bin/arm-linux-androideabi-gcc -- works
(... ... ...)
-- Looking for strnlen
-- Looking for strnlen - not found
-- Looking for snprintf
-- Looking for snprintf - not found
-- Looking for _snprintf
-- Looking for _snprintf - not found
CMake Error at CMakeLists.txt:591 (MESSAGE):
  No snprintf function found, please report!

> OpenALSoft: 1.19.1
> Android NDK r14b

**Comment by kcat** [comment]
I believe the Andoid NDK comes with its own CMake toolchain nowadays. In a fresh build directory, try something like:
```
$ cmake .. -DCMAKE_TOOLCHAIN_FILE=/opt/android-ndk/build/cmake/android.toolchain.cmake
```

I'll probably be removing the XCompile-Android.txt toolchain and just give instructions for using the NDK's toolchain, since it more correctly handles setting up CMake as it expects to be for Android cross-compiling.

If that doesn't work, can you paste the full CMake command line and output from a fresh build directory?

**Comment by LordOfDragons** [comment]
I'm using this as dependent build. My build system by itself is SCons. It simply triggers the OpenAL cmake build internally. I've got the CC/CXX and such values set manually to match the toolchain choosen. So best would be if this is just command line parameters like "-DCMAKE_X=Y" to get it running. I would like not to force SCons finding some cmake files in the NDK.

**Comment by LordOfDragons** [comment]
I gave your command line a test but it is not working:

-- Configuring done
-- Generating done
-- Build files have been written to: /home/roland/progs/deandroid/extern/openal/build/openal-soft-1.19.1/build
Scanning dependencies of target common
[  1%] Building C object CMakeFiles/common.dir/common/alcomplex.c.o
In file included from /home/roland/progs/deandroid/extern/openal/build/openal-soft-1.19.1/common/alcomplex.c:5:
/home/roland/progs/deandroid/extern/openal/build/openal-soft-1.19.1/common/math_defs.h:36:21: error: static declaration of 'log2f' follows non-static
      declaration
static inline float log2f(float f)
                    ^
/opt/roland/android-ndk/toolchains/llvm/prebuilt/linux-x86_64/sysroot/usr/local/include/math.h:38:15: note: previous declaration is here
float         log2f(float);
              ^
1 error generated.
make[2]: *** [CMakeFiles/common.dir/build.make:63: CMakeFiles/common.dir/common/alcomplex.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:68: CMakeFiles/common.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
scons: *** [extern/openal/build/lib/libopenal.a] Error 1
scons: building terminated because of errors.

**Comment by kcat** [comment]
> My build system by itself is SCons. It simply triggers the OpenAL cmake build internally. I've got the CC/CXX and such values set manually to match the toolchain choosen. So best would be if this is just command line parameters like "-DCMAKE_X=Y" to get it running. I would like not to force SCons finding some cmake files in the NDK.

The location of the NDK needs to be known anyway, since that's where the cross-compiler and everything is. The toolchain file is in a fixed location from the NDK's root directory, so once you have that, you're fine. Whether it's a parameter to your SCons configuration, or some kind of `NDK_ROOT` environment variable, you just need to pass `-DCMAKE_TOOLCHAIN_FILE={ndk_root}/build/cmake/android.toolchain.cmake` to cmake.

Also, you shouldn't set the `CC`/`CXX` environment variables to the cross-compiler. That will cause the OpenAL Soft build to fail when it needs to generate native (non-cross-compiled) tools to run during the build.

> I gave your command line a test but it is not working

Can you paste the full output from a fresh build directory? I've seen the issue before of `log2f` not being detected during cmake, but then existing when doing the build, and I can't remember why that happens.

**Comment by LordOfDragons** [comment]
I've experimented with this now a bit. I've removed now the manual cross-compiler set up and put in just the toolchain. This resulted again in the log2f problem. I then noticed a small output in the compile steps that I missed earlier. OpenAL complained about api-level not being set and that it assumes 16. I then went with this setup:

-DANDROID_PLATFORM=18
-DCMAKE_TOOLCHAIN_FILE={NDK_ROOT}/build/cmake/android.toolchain.cmake

This way it works. Looks like OpenAL can not be compiled below api level 18.
